### PR TITLE
feat: more selector types

### PR DESCRIPTION
**Pull request by DeDiamondPro**
Adds support for user, role and channel selector types.

### CODE REVIEW

**Comment by gdude2002** [comment]
Apologies for taking so long to get to this - would you mind checking out that conflict? I should be able to get this merged pretty soon.

**Comment by DeDiamondPro** [comment]
> Apologies for taking so long to get to this - would you mind checking out that conflict? I should be able to get this merged pretty soon.

I'll fix the conflict when I have time tomorrow.

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/_Functions.kt
Comment (gdude2002) [inline kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/_Functions.kt:128]: To ensure that bots don't suddenly stop working, I'd prefer we kept these functions around and deprecated them with a compiler error. This means that already-compiled bots won't break on it if the dependency in their environment is updated.

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/EphemeralSelectMenu.kt
Comment (gdude2002) [inline kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/EphemeralSelectMenu.kt]: This is definitely still a class, not an interface. I don't think this doc comment needed changing at all, verboseness is good.

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/EphemeralSelectMenu.kt
Comment (gdude2002) [inline kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/EphemeralSelectMenu.kt:25]: If `C` is meant to be some kind of context object, then you should give it a base type here.

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/PublicSelectMenu.kt
Comment (gdude2002) [inline kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/PublicSelectMenu.kt]: This really didn't need rewriting, bud. Let's not remove verboseness from existing comments.

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/PublicSelectMenu.kt
Comment (gdude2002) [inline kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/PublicSelectMenu.kt:26]: Again, if `C` is a context type, you need to provide a base type for it. Don't leave generics untyped if at all possible.

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/string/StringSelectMenu.kt
Comment (gdude2002) [inline kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/string/StringSelectMenu.kt:45]: These constants (e.g. `DESCRIPTION_MAX`) should be defined in the file they're used in. I think you forgot to move them over.

File: test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt
Comment (gdude2002) [inline test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt:37]: This seems like it should be a default, no? Why do you specify it here?

If it's not a default, it's probably time to make it so.

File: test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt
Comment (gdude2002) [inline test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt]: You shouldn't need `runBlocking` here or in any of the other action blocks. Isn't `joinToString` inline? Can you investigate why this is?

If there's nothing we can do about this, please store the result in a variable in the `action` block and make use of it in the `respond` block. Don't use `runBlocking`.

File: test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt
Comment (gdude2002) [inline test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt:95]: See the comment I left on the other set of components.

**Review by gdude2002 — CHANGES_REQUESTED** [review]
Overall, I think this is pretty good - just a few things to address. Good work!

File: test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt
Comment (DeDiamondPro) [reply test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt:37]: The default is a maximum of 1 choice, which is the same as it was before. Setting it to null makes max choices the same as the numbers of options you have (no maximum). doing `maximumChoices = 25` would serve the same purpose

Comment (gdude2002) [reply test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/SelectorTestExtension.kt:37]: Oh, hmm, okay, that's fair enough then.

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/EphemeralSelectMenu.kt
Comment (DeDiamondPro) [reply kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/EphemeralSelectMenu.kt:25]: Since `C` needs to meet 2 conditions, it is defined in the where clause a few lines below (`where C : SelectMenuContext, C : EphemeralInteractionContext`)

File: kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/PublicSelectMenu.kt
Comment (DeDiamondPro) [reply kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/components/menus/PublicSelectMenu.kt:26]: Since `C` needs to meet 2 conditions, it is defined in the where clause a few lines below (`where C : SelectMenuContext, C : PublicInteractionContext`)

**Review by gdude2002 — DISMISSED** [review]
This is looking good - thanks for being patient. I think we can merge this once you've merged in the latest changes and verified it all works.

**Comment by gdude2002** [comment]
Looks good, thanks for your work!

**Pull request merged**
